docs/advanced-usage/enabling-natural-language-search.md

```diff
@@ -0,0 +1,129 @@
+
+    :param locators_and_images: An iterator through PIL Image files and their corresponding locators (as provided to
+        the Kolena platform).
+    :param batch_size: Batch size for number of images to extract embeddings for simultaneously. Defaults to 50 to
+        avoid having too many file handlers open at once.
+    ""
+```
+
+### Step 2: Load Images for Extraction
+
+In order to extract embeddings on image data, we must load our image files into a [`PIL.Image.Image`](https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Image) object.
+In this section, we will load these images from an S3 bucket. For other cloud storage services, please refer to your cloud storage's API docs.
+
+```python
+from typing import Iterator
+from typing import Tuple
+
+import boto3
+from urllib.parse import urlparse
+from PIL import Image
+
+s3 = boto3.client("s3")
+
+def load_image_from_locator(locator: str) -> Image.Image:
+    parsed_url = urlparse(locator)
+    bucket_name = parsed_url.netloc
+    key = parsed_url.path.lstrip("/")
+    file_stream = boto3.resource("s3").Bucket(bucket_name).Object(key).get()["Body"]
+    return Image.open(file_stream)
+
+def iter_image_locators(locators: List[str]) -> Iterator[Tuple[str, Image.Image]]:
+    for locator in locators:
+        image = load_image_from_locator(locator)
+        yield locator, image
+```
+
+!!! tip end
+    When processing large scales of images, we recommend using an `Iterator` to limit the number
+    of images loaded into memory at once.
+
+### Step 3: Extract and Upload Embeddings
+
+We first [retrieve and set](https://docs.kolena.io/installing-kolena/#initialization) our `KOLENA_TOKEN` environment variable.
+This is used by the uploader for authentication against your Kolena instance.
+
+```shell
+export KOLENA_TOKEN="********"
+```
+
+We then pass our locators into the `extract_and_upload_embeddings` function to iteratively upload embeddings for all
+[`Image`][kolena.workflow.Image] objects in the Kolena platform with matching locators.
+
+```python
+from kclip.util import extract_and_upload_embeddings
+
+locators = [
+    "s3://kolena-public-datasets/labeled-faces-in-the-wild/imgs/AJ_Cook/AJ_Cook_0001.jpg",
+    "s3://kolena-public-datasets/labeled-faces-in-the-wild/imgs/AJ_Lamas/AJ_Lamas_0001.jpg",
+    "s3://kolena-public-datasets/labeled-faces-in-the-wild/imgs/Aaron_Eckhart/Aaron_Eckhart_0001.jpg",
+]
+extract_and_upload_embeddings(iter_image_locators(locators))
+```
+
+Once the upload completes, we can now visit [<nobr>:kolena-studio-16: Studio</nobr>](https://app.kolena.io/redirect/studio)
+to search by natural language over the corresponding [`Image`][kolena.workflow.Image] data.
+
+## Conclusion
+In this tutorial, we learned how to extract and upload vector embeddings over your [`Image`][kolena.workflow.Image] data.
```

Comment:
We don't actually have a guide on how to use NLS once the embeddings are uploaded. Does it seem like we need to add one? That feels like a whole different beast to document...